mtags/src/main/scala-3/scala/meta/internal/pc/IndexedContext.scala

```diff
@@ -0,0 +1,213 @@
+
+    def findSymbol(name: String): Option[List[Symbol]] = {
+      symbolByName(name).orElse(outer.findSymbol(name))
+    }
+
+    def symbolByName(name: String): Option[List[Symbol]] =
+      names.symbols.get(name).map(_.toList)
+
+    def scopeSymbols: List[Symbol] =
+      names.symbols.values.toList.flatten ++ outers.flatMap(_.scopeSymbols)
+
+    def rename(name: SimpleName): Option[String] = {
+      names.renames
+        .get(name)
+        .orElse(outer.rename(name))
+    }
+
+    private def outers: List[IndexedContext] = {
+      val builder = List.newBuilder[IndexedContext]
+      var curr = outer
+      while (!curr.isEmpty) {
+        builder += curr
+        curr = curr.outer
+      }
+      builder.result
+    }
+
+  }
+
+  enum Result {
+    case InScope, Conflict, Missing
+    def exists: Boolean = this match {
+      case InScope | Conflict => true
+      case Missing => false
+    }
+  }
+
+  private case class Names(
+      symbols: Map[String, List[Symbol]],
+      renames: Map[SimpleName, String]
+  )
+
+  private def extractNames(ctx: Context): Names = {
+
+    def accessibleSymbols(site: Type, tpe: Type)(using
+        Context
+    ): List[Symbol] = {
+      tpe.decls.toList.filter(sym =>
+        sym.isAccessibleFrom(site, superAccess = false)
+      )
+    }
+
+    def accesibleMembers(site: Type)(using Context): List[Symbol] =
+      site.allMembers
+        .filter(denot =>
+          denot.symbol.isAccessibleFrom(site, superAccess = false)
+        )
+        .map(_.symbol)
+        .toList
+
+    def allAccessibleSymbols(
+        tpe: Type,
+        filter: Symbol => Boolean = _ => true
+    )(using Context): List[Symbol] = {
+      val initial = accessibleSymbols(tpe, tpe).filter(filter)
+      val fromPackageObjects =
+        initial
+          .filter(_.isPackageObject)
+          .flatMap(sym => accessibleSymbols(tpe, sym.thisType))
+      initial ++ fromPackageObjects
+    }
+
+    def fromImport(site: Type, name: Name)(using Context): List[Symbol] = {
+      List(site.member(name.toTypeName), site.member(name.toTermName))
+        .flatMap(_.alternatives)
+        .map(_.symbol)
+    }
+
+    def fromImportInfo(
+        imp: ImportInfo
+    )(using Context): List[(Symbol, Option[TermName])] = {
+      val excludedNames = imp.excluded.map(_.decoded)
+
+      if (imp.isWildcardImport) {
+        allAccessibleSymbols(
+          imp.site,
+          sym => !excludedNames.contains(sym.name.decoded)
+        ).map((_, None))
+      } else {
+        imp.forwardMapping.toList.flatMap { (name, rename) =>
+          val isRename = name != rename
+          if (!isRename && !excludedNames.contains(name.decoded))
+            fromImport(imp.site, name).map((_, None))
+          else if (isRename)
+            fromImport(imp.site, name).map((_, Some(rename)))
+          else Nil
+        }
+      }
+    }
+
+    given Context = ctx
+    val (symbols, renames) =
```

Comment:
Aren't we doing almost exactly the same in `NamesInScope` ? Are we able to merge them both? Or reuse one in the other?

Reply:
I forgot to remove `NamesInScope`.  I switched everything on `IndexedContext` usage